File: logic/Story.py
from enum import Enum

import SoundEffect
from logic.Recap import Recap
from logic.Effects import Quality


# TODO Remove source
# How an act was added to the story
# Played from hand, sprung from hand, etc
class Source(Enum):
    HAND = 0
    SPRING = 1
    PILE = 2


class Story:
    def __init__(self):
        self.acts = []
        self.recap = Recap()

    def add_act(self, card, owner, source=Source.HAND, i=None):
        act = Act(card, owner, source)
        if i is None:
            self.acts.append(act)
        else:
            self.acts.insert(i, act)

    def run(self, game, isSimplified=False):
        # Reset the recap so that it now recaps this run
        self.recap.reset()

        state_before_play = ['', '']
        for player in [0, 1]:
            state_before_play[player] = game.get_client_model(player=player, is_recap=True)
        self.recap.add_state(state_before_play)
        game.animations = [[], []]

        index = 0
        # List of callbacks that occur as the round ends
        round_end_effects = []
        while self.acts:
            act = self.acts.pop(0)

            game.sound_effect = SoundEffect.Resolve

            if act.countered:
                result = 'Countered'
            else:
                # Only for the tutorial
                if isSimplified:
                    game.score[act.owner] += act.card.points
                    result = 'SIMPLIFIED TODO'

                elif act.source is Source.HAND or act.source is Source.PILE:
                    result = act.card.play(player=act.owner,
                                           game=game,
                                           index=index,
                                           bonus=act.bonus)
                    round_end_effects.append((act.card.on_round_end, act.owner))

            # Card goes to pile unless it has fleeting
            if Quality.FLEETING not in act.card.qualities:
                game.pile[act.owner].append(act.card)
            else:
                game.expended[act.owner].append(act.card)

            index += 1

            self.recap.add(act.card, act.owner, result)

            # Save as a string what state the game is in after the card has been played
            state_after_play = ['', '']
            for player in [0, 1]:
                state_after_play[player] = game.get_client_model(player=player, is_recap=True)

            self.recap.add_state(state_after_play)
            game.animations = [[], []]

        # Do any round end effects
        for (callback, player) in round_end_effects:
            callback(player, game)

    def save_end_state(self, game):
        # Save an ending state which includes win/loss/tie sfx
        state_after_play = ['', '']
        for player in [0, 1]:
            # Win/Lose based on which player the state is for
            if self.recap.wins[player] > 0:
                game.sound_effect = SoundEffect.Win
            elif self.recap.wins[player ^ 1] > 0:
                game.sound_effect = SoundEffect.Lose
            else:
                game.sound_effect = SoundEffect.Tie

            state_after_play[player] = game.get_client_model(player=player, is_recap=True)
        self.recap.add_state(state_after_play)
        game.animations = [[], []]

    def clear(self):
        self.acts = []

    def get_length(self):
        return len(self.acts)

    def is_empty(self):
        return len(self.acts) == 0

    def counter(self, function):
        for act in self.acts:
            if function(act):
                act.countered = True

                return act.card

        return None
        # if self.get_length() > index:
        #     self.acts[index].countered = True
        #
        #     return self.acts[index].card
        # else:
        #     return None

    # TODO This has a bug if origin is before destination, since the indexing changes in that case
    def move_act(self, index_origin, index_dest):
        act = self.acts.pop(index_origin)
        self.acts.insert(index_dest, act)

        return act

    # Remeove and return the act at index
    def remove_act(self, index):
        if len(self.acts) < index + 1:
            raise Exception(f"Tried to remove act {index} in a story with only {len(self.acts)} acts.")

        return self.acts.pop(index)

    # Replace the act at index with replacement_act
    def replace_act(self, index, replacement_act):
        if len(self.acts) < index + 1:
            raise Exception(f"Tried to replace act {index} in a story with only {len(self.acts)} acts.")

        self.acts[index] = replacement_act


class Act:
    def __init__(self, card, owner, source=Source.HAND):
        self.card = card
        self.owner = owner
        self.source = source

        self.countered = False
        self.bonus = 0

File: shared/state/story.ts
import Card from '../../shared/state/card'

import { SoundEffect } from './soundEffect'
import type GameModel from './gameModel'
import getClientGameModel from './clientGameModel'
import Act from './act'

class Story {
  acts: Act[] = []

  // The index in the story of the current act
  currentIndex: number = 0

  // Add a card to the story with given owner and at given position
  addAct(card: Card, owner: number, i?: number) {
    const act = new Act(card, owner)
    if (i === undefined) {
      this.acts.push(act)
    } else {
      this.acts.splice(i, 0, act)
    }
  }

  // Run the current story
  run(game: GameModel) {
    game.score = [0, 0]
    game.recentModels = [[], []]

    // Add a model at the start
    game.versionIncr()
    addRecentModels(game)

    const roundEndEffects: [Function, number][] = []
    for (
      this.currentIndex = 0;
      this.currentIndex < this.acts.length;
      this.currentIndex++
    ) {
      const act = this.acts[this.currentIndex]

      // Reset the sound, changed if the card sets it
      game.sound = SoundEffect.Resolve
      act.card.play(act.owner, game, this.currentIndex, 0)

      // Add any card end effects the card may have
      // TODO Just scan at the end before discarding the card
      roundEndEffects.push([act.card.onRoundEndIfThisResolved, act.owner])

      addRecentModels(game)
    }

    // Do all round end effects
    for (const [callback, player] of roundEndEffects) {
      callback(player, game)
    }
  }

  // Save the final state of the story resolving, and clear the story
  saveFinalStateAndClear(game: GameModel) {
    addRecentModels(game)

    // Set winner/loser/tie sfx
    if (game.score[0] > game.score[1]) {
      game.recentModels[0][game.recentModels[0].length - 1].sound =
        SoundEffect.Win
      game.recentModels[1][game.recentModels[1].length - 1].sound =
        SoundEffect.Lose
    } else if (game.score[0] < game.score[1]) {
      game.recentModels[0][game.recentModels[0].length - 1].sound =
        SoundEffect.Lose
      game.recentModels[1][game.recentModels[1].length - 1].sound =
        SoundEffect.Win
    } else {
      game.recentModels[0][game.recentModels[0].length - 1].sound =
        SoundEffect.Tie
      game.recentModels[1][game.recentModels[1].length - 1].sound =
        SoundEffect.Tie
    }

    // Clear the story
    this.acts = []
  }

  // Remove the act at the given index
  removeAct(index: number): Act {
    if (this.acts.length <= index) {
      throw new Error(
        `Tried to remove act ${index} in a story with only ${this.acts.length} acts.`,
      )
    }

    return this.acts.splice(index, 1)[0]
  }

  // Replace the act at the given index with the given act
  replaceAct(index: number, act: Act) {
    if (this.acts.length <= index) {
      throw new Error(
        `Tried to replace act ${index} in a story with only ${this.acts.length} acts.`,
      )
    }

    this.acts[index] = act
  }

  // Return a full deepcopy of the story
  getDeepCopy(): Story {
    let copy = new Story()

    this.acts.forEach((act) => {
      copy.acts.push({ ...act })
    })

    copy.currentIndex = this.currentIndex

    return copy
  }
}

// Add the current state to list of remembered recent states
function addRecentModels(model: GameModel): void {
  // Get a recent model for each and add for that player
  const model0 = getClientGameModel(model, 0, true)
  model0.recentModels = [[], []]
  model0.isRecap = true
  model.recentModels[0].push(model0)

  const model1 = getClientGameModel(model, 1, true)
  model1.recentModels = [[], []]
  model1.isRecap = true
  model.recentModels[1].push(model1)

  // Increment the version
  model.versionIncr()
}

export { Act, Story }
